src/rrd_open.c

```diff
@@ -599,7 +609,72 @@ int rrd_lock(
     return rrd_rwlock(rrd_file, 1);
 }
 
-static int rrd_rwlock(rrd_file_t *rrd_file, int writelock)
+#if defined(_WIN32) && !defined(__CYGWIN__) && !defined(__CYGWIN32__)
+#define USE_WINDOWS_LOCK 1
+#endif
+
+#ifdef USE_WINDOWS_LOCK
+static
+int rrd_windows_lock(
+    int fd)
+{
+    int ret;
+
+    while (1) {
+        ret = _locking(fd, _LK_NBLCK, LONG_MAX);
+        if (ret == 0)
+            break; /* success */
+        if (errno != EACCES)
+            break; /* failure */
+        /* EACCES: someone else has the lock. */
+
+        /*
+         * Wait 0.01 seconds before trying again.  _locking()
+         * with _LK_LOCK would work similarly but waits 1 second
+         * between tries, which seems less desirable.
+         */
+        Sleep(10);
+    }
+
+    return ret;
+}
+#endif
+
+static
+int close_and_unlock(
+    int fd)
+{
+    int ret;
```

Comment:
I noticed this ret needs initializing.  Not to spread the blame, but I'm surprised the compiler doesn't warn about it.  It warned me when I had the same problem in rrd_close().